---
title: "Class18 - Pertussis"
author: "Peter"
format: html
toc: true
---

## Background

Pertussis (a.k.a. Whooping cough) is a common lung infection caused by the bacteria *B. Pertussis*.

The CDC tracks cases of Pertussis in the US

## Investigating Pertussis cases by year

```{r}
library(datapasta)
```

```{r, include=FALSE}
cdc <- data.frame(
                                 Year = c(1922L,1923L,1924L,1925L,
                                          1926L,1927L,1928L,1929L,1930L,1931L,
                                          1932L,1933L,1934L,1935L,1936L,
                                          1937L,1938L,1939L,1940L,1941L,1942L,
                                          1943L,1944L,1945L,1946L,1947L,
                                          1948L,1949L,1950L,1951L,1952L,
                                          1953L,1954L,1955L,1956L,1957L,1958L,
                                          1959L,1960L,1961L,1962L,1963L,
                                          1964L,1965L,1966L,1967L,1968L,1969L,
                                          1970L,1971L,1972L,1973L,1974L,
                                          1975L,1976L,1977L,1978L,1979L,1980L,
                                          1981L,1982L,1983L,1984L,1985L,
                                          1986L,1987L,1988L,1989L,1990L,
                                          1991L,1992L,1993L,1994L,1995L,1996L,
                                          1997L,1998L,1999L,2000L,2001L,
                                          2002L,2003L,2004L,2005L,2006L,2007L,
                                          2008L,2009L,2010L,2011L,2012L,
                                          2013L,2014L,2015L,2016L,2017L,2018L,
                                          2019L,2020L,2021L,2022L,2023L,2024L),
                                  Cases = c(107473,164191,165418,152003,
                                          202210,181411,161799,197371,
                                          166914,172559,215343,179135,265269,
                                          180518,147237,214652,227319,103188,
                                          183866,222202,191383,191890,109873,
                                          133792,109860,156517,74715,69479,
                                          120718,68687,45030,37129,60886,
                                          62786,31732,28295,32148,40005,
                                          14809,11468,17749,17135,13005,6799,
                                          7717,9718,4810,3285,4249,3036,
                                          3287,1759,2402,1738,1010,2177,2063,
                                          1623,1730,1248,1895,2463,2276,
                                          3589,4195,2823,3450,4157,4570,
                                          2719,4083,6586,4617,5137,7796,6564,
                                          7405,7298,7867,7580,9771,11647,
                                          25827,25616,15632,10454,13278,
                                          16858,27550,18719,48277,28639,32971,
                                          20762,17972,18975,15609,18617,
                                          6124,2116,3044,7063,35493)
       )
```

> Q1. Plot the number of cases per year using ggplot

```{r}
library(ggplot2)
plot <- ggplot(cdc) + 
  aes(Year, Cases) +
  geom_point() + 
  geom_line() +
  labs(title="Pertussis Cases By Year", x="Years", y="Number of Cases")
plot
```

> Q2. Add some key time points in the history of the pertussis vaccine. Include the wP roll out (first vaccine) in 1946 and the switch to aP in 1996.


```{r}
plot + 
  geom_vline(xintercept=1946, col="blue") +
  geom_vline(xintercept=1996, col="red") +
  geom_vline(xintercept=2020, col="purple")
```

After the wP vaccine, cases decreased dramatically. Cases remained stagnant after the switch to the aP vaccine. There were some outbreaks during the early 2010s, but case numbers dropped during COVID-19. The current trends suggest Pertussis is making a bit of a comeback.


> Q3. Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?

Trends suggest that switching to the aP vaccine reduced vaccine hesitancy and it provided protection inferior to the wP vaccine. The change was originally made due to side effects of the wP vaccine, but this came with the consequence of potential reduced efficacy.


## Exploring CMI-PB data

CMI-PB stands for Computational Models of Immunity - Pertussis Boost. Its major goal is to investigate how the immune system responds differently with the aP or wP vaccine.

```{r}
library(jsonlite)
subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE)
head(subject)
```

> Q4. How many aP and wP infancy vaccinated subjects are in the dataset?

```{r}
table(subject$infancy_vac)
```

> Q5. How many Male and Female subjects/patients are in the dataset?

```{r}
table(subject$biological_sex)
```


> Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?

```{r}
table(subject$race, subject$biological_sex)
```
### Using lubridate

```{r}
library(lubridate)
today() - ymd("2000-01-01")
time_length( today() - ymd("2000-01-01"),  "years")
```

> Q7. Using this approach determine (i) the average age of wP individuals, (ii) the average age of aP individuals; and (iii) are they significantly different?

(i) 
```{r}
library(dplyr)
subject$age <- today() - ymd(subject$year_of_birth)
wP <- subject %>% filter(infancy_vac == "wP")
round( summary( time_length( wP$age, "years" ) ) )
```
(ii)
```{r}
subject$age <- today() - ymd(subject$year_of_birth)
aP <- subject %>% filter(infancy_vac == "aP")
round( summary( time_length( aP$age, "years" ) ) )
```

(iii)
```{r}
p <- t.test(time_length(aP$age, "years"), time_length(wP$age, "years"))
p$p.value
```



> Q8. Determine the age of all individuals at time of boost.

```{r}
int <- ymd(subject$date_of_boost) - ymd(subject$year_of_birth)
age_at_boost <- time_length(int, "year")
head(age_at_boost)
```
> Q9. With the help of a faceted boxplot or histogram (see below), do you think these two groups are significantly different?

```{r}
ggplot(subject) +
  aes(time_length(age, "year"),
      fill=as.factor(infancy_vac)) +
  geom_histogram(show.legend=FALSE) +
  facet_wrap(vars(infancy_vac), nrow=2) +
  xlab("Age in years")
```
They appear to be very different, and the p-value we got in question 7 (iii) is statistically significant.


### Joining multiple tables

```{r}
specimen <- read_json("http://cmi-pb.org/api/v5_1/specimen", simplifyVector = TRUE) 
titer <- read_json("https://www.cmi-pb.org/api/v5_1/plasma_ab_titer", simplifyVector = TRUE)
```

> Q9. Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details:

```{r}
meta <- inner_join(subject, specimen)
dim(meta)
head(meta)
```

> Q10. Now using the same procedure join meta with titer data so we can further analyze this data in terms of time of visit aP/wP, male/female etc

```{r}
abdata <- inner_join(titer, meta)
head(abdata)
```

> Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?

```{r}
dim(abdata)
```
```{r}
table(abdata$isotype)
```

> Q12. What are the different $dataset values in abdata and what do you notice about the number of rows for the most “recent” dataset?

```{r}
table(abdata$dataset)
```
There are more than double the amount of the previous year...


> Make a boxplot of antigen levels across the whole dataset (MFI vs antigen).

```{r}
ggplot(abdata) + 
  aes(MFI, antigen) + 
  geom_boxplot()
```

## Examine the IgG Ab titer levels

```{r}
igg <- abdata %>% filter(isotype == "IgG")
head(igg)
```

> Q13. Complete the following code to make a summary boxplot of Ab titer levels (MFI) for all antigens:

```{r}
ggplot(igg) +
  aes(MFI_normalised, antigen) +
  geom_boxplot() + 
    xlim(0,75) +
  facet_wrap(vars(visit), nrow=2)
```

> Q14. What antigens show differences in the level of IgG antibody titers recognizing them over time? Why these and not others?

Antigens PT, PRN, FIM2/3, and FHA show considerable differences over time. Thus could be due to the fact that they are antigens specific to Pertussis, while DT and TT are present in the DTAP vaccine for Diphtheria and Tetanus, not pertussis necessarily


```{r}
ggplot(igg) +
  aes(MFI_normalised, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(visit), nrow=2) +
  xlim(0,75) +
  theme_bw()
```

```{r}
igg %>% filter(visit != 8) %>%
ggplot() +
  aes(MFI_normalised, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  xlim(0,75) +
  facet_wrap(vars(infancy_vac, visit), nrow=2)
```

> Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a “control” antigen (“OVA”, that is not in our vaccines) and a clear antigen of interest (“PT”, Pertussis Toxin, one of the key virulence factors produced by the bacterium B. pertussis).

```{r}
filter(igg, antigen=="OVA") %>%
  ggplot() +
  aes(MFI_normalised, col=infancy_vac) +
  geom_boxplot(show.legend = TRUE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

```{r}
filter(igg, antigen=="FHA") %>%
  ggplot() +
  aes(MFI_normalised, col=infancy_vac) +
  geom_boxplot(show.legend = TRUE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

> Q16. What do you notice about these two antigens time courses and the PT data in particular?

The OVA time course is less interesting than that of the PT time course, where at the 8th time point the aP values were replaced with wP such that the entire plot is wP. I personally compared OVA to FHA, where this phenomenon was not observed.

> Q17. Do you see any clear difference in aP vs. wP responses?

Besides the strange data at time point 8, there does not appear to be signicant difference in the time trajectory of wP vs aP.

```{r}
abdata.21 <- abdata %>% filter(dataset == "2021_dataset")

abdata.21 %>% 
  filter(isotype == "IgG",  antigen == "PT") %>%
  ggplot() +
    aes(x=planned_day_relative_to_boost,
        y=MFI_normalised,
        col=infancy_vac,
        group=subject_id) +
    geom_point() +
    geom_line() +
    geom_vline(xintercept=0, linetype="dashed") +
    geom_vline(xintercept=14, linetype="dashed") +
  labs(title="2021 dataset IgG PT",
       subtitle = "Dashed lines indicate day 0 (pre-boost) and 14 (apparent peak levels)")
```
```{r}
abdata.20 <- abdata %>% filter(dataset == "2020_dataset")

abdata.20 %>% 
  filter(isotype == "IgG",  antigen == "PT") %>%
  ggplot() +
    aes(x=planned_day_relative_to_boost,
        y=MFI_normalised,
        col=infancy_vac,
        group=subject_id) +
    geom_point() +
    geom_line() +
    geom_vline(xintercept=0, linetype="dashed") +
    geom_vline(xintercept=14, linetype="dashed") +
  labs(title="2020 dataset IgG PT",
       subtitle = "Dashed lines indicate day 0 (pre-boost) and 14 (apparent peak levels)")
```

> Q18. Does this trend look similar for the 2020 dataset?

Absolutely not. The 2020 dataset sees an abrupt change around day 115ish where values for aP halt and wP values slowly approach 0 over a longer period of time. The behavior of aP in particular is much different in the 2021 dataset where it is more gradual.


## Obtaining CMI-PB RNASeq Data

For RNA-Seq data the API query mechanism quickly hits the web browser interface limit for file size. We will present alternative download mechanisms for larger CMI-PB datasets in the next section. However, we can still do “targeted” RNA-Seq querys via the web accessible API.

```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versione>d_ensembl_gene_id=eq.ENSG00000211896.7"
#rna <- read_json(url, simplifyVector = TRUE) 
```

To facilitate further analysis we need to “join” the rna expression data with our metadata meta, which is itself a join of sample and specimen data. This will allow us to look at this genes TPM expression values over aP/wP status and at different visits (i.e. times):

```{r}
#meta <- inner_join(specimen, subject)
#ssrna <- inner_join(rna, meta)
```

> Q19. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).

```{r}
#ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```

> Q20. What do you notice about the expression of this gene (i.e. when is it at it’s maximum level)?

It trends higher towards the middle rather than peaking initially or gradually rising to a larger peak as time progresses

> Q21. Does this pattern in time match the trend of antibody titer data? If not, why not?

It does especially for PT which peaked around time point 5 according to the previous boxplot data

```{r}
#ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))
```
```{r}
#ssrna %>%  
  filter(visit==4) %>% 
  #ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug() 
```

**I was originally able to plot all of the graphs (specifically the one for Q19) but the internet connection would not render them so I just hashed all of the code that I used. I was able to analyze the plots easily because they were identical to the ones on the lab sheet**



















